### PR TITLE
Teach the CLI to lazily load commands

### PR DESCRIPTION
**Pull request by rictic**
The command is resolved and loaded from the cwd, and if it isn't installed we offer to install it for the user.

### CODE REVIEW

File: packages/labs/cli/src/lib/command.ts [hunk 1]
```diff
@@ -48,9 +60,65 @@ export interface Command {
...
+  name: string;
+  description: string;
+  aliases?: string[];
+  /** Note: not guaranteed to be complete! */
+  subcommands?: Command[];
+  /**
+   * The import specifier that resolves to an ES module whose exports match
+   * the CommandModule interface.
+   */
+  importSpecifier: string;
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/command.ts:78]: If it just does a dynamic import the module resolution will be based on where the command is installed, which might be global. We want to resolve relative to the user's cwd

File: packages/labs/cli/src/lib/lit-cli.ts [hunk 2]
```diff
@@ -112,17 +131,31 @@ export class LitCli {
...
+    | {invalidCommand: string}
+    | {commandNotInstalled: boolean}
+  > {
     try {
       const parsedArgs = commandLineCommands([...commands.keys()], [...args]);
-      const commandName = parsedArgs.command!;
-      const command = commands.get(commandName)!;
+      const commandName = parsedArgs.command;
+      let command = commandName && commands.get(commandName);
+      if (!commandName || command == null || command == '') {
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/lit-cli.ts:147]: Different variables, `commandName` vs `command`

File: packages/labs/cli/package.json [hunk 3]
```diff
@@ -13,10 +13,44 @@
...
+  },
+  "wireit": {
+    "build": {
+      "#TODO": "this needs to depend on ../../localize-tools:build I think",
+      "files": [
+        "tsconfig.json",
+        "src/**/*"
+      ],
+      "#comment": "This never fails and always emits output so that we can run tests of code that doesn't type check",
+      "command": "tsc --skipLibCheck >/dev/null || echo ''"
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/package.json]: +1, done

File: packages/labs/cli/src/lib/command.ts [hunk 4]
```diff
@@ -48,9 +60,65 @@ export interface Command {
...
+ * This is the type that an ES module that a ReferenceToCommand refers to should
+ * obey.
+ */
+export interface CommandModule {
+  /**
+   * Note that this can return a ReferenceToCommand, which the CLI will
+   * further resolve. This is generally preferable to attempting to do
+   * reference resolution yourself.
+   */
+  getCommand(opts: GetCommandsOptions): Promise<Command>;
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/command.ts:122]: Yeah, in almost all cases you're right, it's just cheap to make it async now, and more expensive to do so later (would mean that a command that needs async would need a minimum version of the CLI that knows to await it)

File: packages/labs/cli/src/lib/command.ts [hunk 5]
```diff
@@ -48,9 +60,65 @@ export interface Command {
...
+
+/**
+ * This is the type that an ES module that a ReferenceToCommand refers to should
+ * obey.
+ */
+export interface CommandModule {
+  /**
+   * Note that this can return a ReferenceToCommand, which the CLI will
+   * further resolve. This is generally preferable to attempting to do
+   * reference resolution yourself.
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/command.ts]: Expanded out the comment, PTAL

File: packages/labs/cli/src/lib/command.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/command.ts:82]: Expanded comment here

File: packages/labs/cli/src/lib/command.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/command.ts]: Updated types to be explicit that it's allowed to be synchronous

File: packages/labs/cli/src/lib/command.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/labs/cli/src/lib/command.ts:82]: thanks!

File: packages/labs/cli/src/lib/lit-cli.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/labs/cli/src/lib/lit-cli.ts:147]: oops!

File: packages/labs/cli/src/test/help_test.ts [hunk 6]
```diff
@@ -4,62 +4,252 @@
...
+  await cli.run();
+  const output = console.outputStream.text;
+  assert.equal(console.errorStream.buffer.length, 0);
+  assert.match(
+    output,
+    /\s+foo\s+This is the description in the `foo` reference./
+  );
+});
+
+test('help includes unresolved external command descriptions', async ({
```
Comment (justinfagnani) [inline packages/labs/cli/src/test/help_test.ts:119]: is this the same as the previous test?

File: packages/labs/cli/src/test/temp-filesystem-rig.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+import type {Stats} from 'fs';
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = pathlib.dirname(__filename);
+const repoRoot = pathlib.resolve(__dirname, '..', '..', '..', '..');
+
+/**
+ * A test rig for managing a temporary filesystem.
+ */
+export class FilesystemTestRig {
```
Comment (justinfagnani) [inline packages/labs/cli/src/test/temp-filesystem-rig.ts:20]: We should move something like this into the internal test package.

**Review by justinfagnani — APPROVED** [review]

File: packages/labs/cli/src/test/temp-filesystem-rig.ts (hunk 7 above)
Comment (rictic) [reply packages/labs/cli/src/test/temp-filesystem-rig.ts:20]: +1, this was extracted from wireit

File: packages/labs/cli/src/test/help_test.ts (hunk 6 above)
Comment (rictic) [reply packages/labs/cli/src/test/help_test.ts:119]: So it is! Removed

File: packages/labs/cli/src/test/help_test.ts [hunk 8]
```diff
@@ -4,62 +4,232 @@
...
+    this.outputStream = outputStream;
+    this.errorStream = errorStream;
+  }
+}
+
+test.before.each(async (ctx) => {
+  ctx.console = new TestConsole();
+  ctx.fs = new FilesystemTestRig();
+  ctx.stdinLines = [];
+  ctx.stdin = stream.Readable.from(ctx.stdinLines);
```
Comment (kevinpschaaf) [inline packages/labs/cli/src/test/help_test.ts:45]: So if I understand how this works, you need to push onto `stdinLines` before the CLI would have done the read from stdin, right?

Comment (rictic) [reply packages/labs/cli/src/test/help_test.ts:45]: Yes exactly, and generally before you start the CLI running

Comment (rictic) [reply packages/labs/cli/src/test/help_test.ts:45]: Added a comment.

File: packages/labs/cli/src/lib/lit-cli.ts [hunk 9]
```diff
@@ -112,17 +131,30 @@ export class LitCli {
...
+        return {invalidCommand: commandName ?? 'unknown comand'};
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/lit-cli.ts:148]: Done

File: packages/labs/cli/src/lib/lit-cli.ts [hunk 10]
```diff
@@ -153,4 +185,148 @@ export class LitCli {
...
+      reference,
+      resolvedPackageLocation
+    );
+    if (command.kind === 'reference') {
+      return this.#resolveCommandAndMaybeInstallNeededDeps(command);
+    }
+    return command;
+  }
+
+  async #installDepWithPermission(
```
[the remark this answers is not in the document]
Comment (rictic) [reply packages/labs/cli/src/lib/lit-cli.ts:270]: Yeah, makes sense. I've been working in wireit world where we're using #private, but for the lit monorepo TypeScript private seems like the better option for now.

File: packages/labs/cli/src/lib/command.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/labs/cli/src/lib/command.ts]: 👍

File: packages/labs/cli/src/test/temp-filesystem-rig.ts (hunk 7 above)
Comment (kevinpschaaf) [reply packages/labs/cli/src/test/temp-filesystem-rig.ts:20]: Added in https://github.com/lit/lit/pull/2875

File: packages/labs/cli/src/test/uvu-wrapper.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
+import * as uvu from 'uvu';
```
Comment (kevinpschaaf) [inline packages/labs/cli/src/test/uvu-wrapper.ts:1]: Can do in a follow-up, but wonder if this is another thing that should maybe move to `tests/src` now that we have a consolidated place to put test utils?

**Review by kevinpschaaf — APPROVED** [review]
LGTM

**Pull request merged**